pkg/services/provisioning/dashboards/file_reader_test.go

```diff
@@ -348,6 +369,23 @@ func (s *fakeDashboardProvisioningService) GetProvisionedDashboardData(name stri
 }
 
 func (s *fakeDashboardProvisioningService) SaveProvisionedDashboard(dto *dashboards.SaveDashboardDTO, provisioning *models.DashboardProvisioning) (*models.Dashboard, error) {
+	// Copy the structs as we need to change them but do not want to alter outside world.
```

Comment:
Quite a few changes to make the mock behave properly. Maybe I am not getting something and should strive for different approach? @marefr suggested using mocking bus calls and as seen in dashboard_service_test.go it definitely looks better (it basically have similar syntax as mocking lib would have) but that would change this to integration test with more interdependencies, as here you would have to match proper bus handlers through 2 services and also not sure if any mocking usecase could be solved with mocking bus calls.

Reply:
Maybe have similar struct to this instead and just increment each field on every operation would be easier?
```go
type fakeDashboardProvisioningService struct {
	inserted     []*dashboards.SaveDashboardDTO		
	provisioned  map[string][]*models.DashboardProvisioning		
	unprovisioned  map[string][]*models.DashboardProvisioning
	deletedProvisioned  map[string][]*models.DashboardProvisioning
	getDashboard []*models.Dashboard		
}	
```

Right now with the mock of `SaveProvisionedDashboard` feels like you try to mock a lot of different things?

Reply:
Yes but coming to the existing mock implementation it seemed like the goal for it was to be internally consistent (and looking at other tests it seemed to be the point of manual mocks). So for example if I do `SaveProvisionedDashboard` I should be able to retrieve it with `GetProvisionedDashboardData` (that is how it was already implemented). So adding `DeleteProvisionedDashboard` to the mock I assumed that if I delete a dashboard I should not be returning it with `GetProvisionedDashboardData` to keep the consistency.

What you propose would definitely be easier, though it would probably need something like `getProvisionedDashboardDataReturnValue` to be able to also mock the getter. On the other hand that is very close to what standard mock libraries do, where you just record what was called and compare it what should have been called.

Thats probably why I am confused with the manual mocking in the codebase and lack of mocking library, and do not know whether to change existing code or just try to extend it.

Reply:
Gotcha. Haven't been involved in the provisioning code earlier so didn't know that we had such a mock.

Reply:
Either extending current or creating a new fake for those cases would be fine. If both feel like to much work I'm fine with manual testing as well.

In general, I think both me and @xlson isn't that happy about the code quality in the provisioning package. 
The fact that we need to write such complicated mocks is a signal that the code might need some rework. Once this is done we could discuss how to improve it.